src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+                np.multiply(
+                    np.multiply(
+                        np.multiply(np.multiply(np.sign(outer_sine), outer_cosine), a),
+                        np.sin(2 * theta),
+                    ),
+                    signal,
+                )
+            )
+
+            derivative_sqrt = np.divide(
+                1, 2 * np.sqrt(np.multiply(sum_A, np.sum(np.power(outer_sine, 2))))
+            )
+            derivative_sine2 = 2 * np.multiply(outer_sine, outer_cosine)
+
+            #  gradients of denomenator
+            dx_c2 = np.multiply(
+                derivative_sqrt,
+                np.multiply(sum_A, np.sum(np.multiply(derivative_sine2, sine_x_term))),
+            )
+            dy_c2 = np.multiply(
+                derivative_sqrt,
+                np.multiply(sum_A, np.sum(np.multiply(derivative_sine2, sine_y_term))),
+            )
+            dz_c2 = np.multiply(
+                derivative_sqrt,
+                np.multiply(sum_A, np.sum(np.multiply(derivative_sine2, sine_z_term))),
+            )
+
+            # gradients
+            dx = np.divide(
+                np.multiply(dx_c1, c2) - np.multiply(dx_c2, c1), np.power(c2, 2)
+            )
+            dy = np.divide(
+                np.multiply(dy_c1, c2) - np.multiply(dy_c2, c1), np.power(c2, 2)
+            )
+            dz = np.divide(
+                np.multiply(dz_c1, c2) - np.multiply(dz_c2, c1), np.power(c2, 2)
+            )
+
+            # update
+            x = x + np.multiply(alpha1, dx)
+            y = y + np.multiply(alpha2, dy)
+            z = z + np.multiply(alpha2, dz)
+
+            if iter_no < 2:
+                stop_cond = np.minimum(np.minimum(dx, dy), dz) / 1000
+
+            if iter_no > 400:
+                stop_cond = np.maximum(np.maximum(dx, dy), dz) + 1
+
+            iter_no = iter_no + 1
+
+        df1 = np.divide(x + np.abs(y + z * 1j), 2)
+        df2 = np.divide(x - np.abs(y + z * 1j), 2)
+        angle_ast = np.angle(y + z * 1j) / 2
+
+        inner_cosine = np.multiply(y, np.cos(2 * theta)) + np.multiply(
+            z, np.sin(2 * theta)
+        )
+        outer_sine = np.sin(np.multiply(a, x) + np.multiply(a, inner_cosine) - b)
+        outer_cosine = np.cos(np.multiply(a, x) + np.multiply(a, inner_cosine) - b)
+
+        sine_x_term = a
+        sine_y_term = np.multiply(a, np.cos(2 * theta))
+        sine_z_term = np.multiply(a, np.sin(2 * theta))
+
+        c1 = np.sum(np.multiply(np.abs(outer_sine), signal))
+        c2 = np.sqrt(np.multiply(sum_A, np.sum(np.power(outer_sine, 2))))
+
+        p = np.divide(c1, c2)
+
+        return df1, df2, angle_ast, p
+
+    def ctf_locate_minima(self, signal):
+        """
+        Find zero-crossings of power spectrum.
+
+        :param signal: Estimated power spectrum
+        :return: array of zero-crossings indices.
+        """
+
+        N = signal.shape[0]
+
+        derivatives = np.zeros((8, signal.shape[0], signal.shape[1]))
```

Comment:
What?

Reply:
@janden , do you not like the syntax finding the differences below here? I know there is np.diff for finding kth difference... if you think that is clearer.  If i'm missing your concern lmk. Thanks

Reply:
I just can't figure out what these lines do. I get that they're supposed to compute some finite differences, but it's not obvious to me.

Reply:
yeah, they are finding first differences.  I'll try to rewrite with np.diff and add a note. I agree so many `slice` instances in there can be a little jarring.

Reply:
>  I'll try to rewrite with np.diff and add a note.

Yeah I think that should make it much clearer.

Reply:
Actually, I was working through this and realized its just dead code. I have removed it. (As I translated the slices to diffs I realized it should be crashing as written, and was not...)

Reply:
That's always a happy surprise…